src/StackExchange.Redis/ConfigurationOptions.cs

```diff
@@ -296,24 +297,61 @@ private static RemoteCertificateValidationCallback TrustIssuerCallback(X509Certi
+                {
+                    return true;
+                }
+                // If we're not valid due to chain errors - check against the trusted issuer
+                // Note that we're only proceeding at all here if the *only* issue is chain errors (not more flags in SslPolicyErrors)
+                return sslPolicyError == SslPolicyErrors.RemoteCertificateChainErrors
+                       && certificate is X509Certificate2 v2
+                       && CheckTrustedIssuer(v2, certificateChain, issuer);
+            };
         }
 
-        private static bool CheckTrustedIssuer(X509Certificate2 certificateToValidate, X509Certificate2 authority)
+        private static bool CheckTrustedIssuer(X509Certificate2 certificateToValidate, X509Chain? chainToValidate, X509Certificate2 authority)
         {
-            // reference: https://stackoverflow.com/questions/6497040/how-do-i-validate-that-a-certificate-was-created-by-a-particular-certification-a
-            X509Chain chain = new X509Chain();
+            // Reference:
+            // https://stackoverflow.com/questions/6497040/how-do-i-validate-that-a-certificate-was-created-by-a-particular-certification-a
+            // https://github.com/stewartadam/dotnet-x509-certificate-verification
+            using X509Chain chain = new X509Chain();
             chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
-            chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
             chain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority;
-            chain.ChainPolicy.VerificationTime = DateTime.Now;
+            chain.ChainPolicy.VerificationTime = chainToValidate?.ChainPolicy?.VerificationTime ?? DateTime.Now;
             chain.ChainPolicy.UrlRetrievalTimeout = new TimeSpan(0, 0, 0);
 
             chain.ChainPolicy.ExtraStore.Add(authority);
-            return chain.Build(certificateToValidate);
+            // This only verifies that the chain is valid, but with AllowUnknownCertificateAuthority could trust
+            // self-signed or partial chained vertificates
+            try
+            {
+                var chainIsVerified = chain.Build(certificateToValidate);
+                if (chainIsVerified)
+                {
+                    // Our method is "TrustIssuer", which means any intermediate cert we're being told to trust
+                    // is a valid thing to trust, up until it's a root CA
+                    bool found = false;
+                    foreach (var chainElement in chain.ChainElements)
+                    {
+                        using var chainCert = chainElement.Certificate;
+                        if (!found && chainCert.RawData.SequenceEqual(authority.RawData))
+                        {
+                            found = true;
+                        }
+                    }
+                    return found;
```

Comment:
The `RawData` property has the behavior of allocating by returning defensive copies. That can't really be avoided for each intermediate, but you could stash the RawData from the authority so it doesn't allocate in each loop.

.NET 8 has RawDataMemory which does not allocate if and when you ever start to target .NET 8, you could consider conditionally using `RawDataMemory`.

```suggestion
                    bool found = false;
                    byte[] authorityBytes = authority.RawData;
                    foreach (var chainElement in chain.ChainElements)
                    {
                        using var chainCert = chainElement.Certificate;
                        if (!found && chainCert.RawData.SequenceEqual(authorityBytes))
                        {
                            found = true;
                        }
                    }
                    return found;
```

Reply:
ACK - no target at the moment but will keep in mind!

Reply:
@NickCraver  maybe we should add a pre-emptive:

``` c#
#if NET8_0_OR_GREATER
#error TODO: use RawDataMemory (needs testing)
#endif
```

??